src/compose/preprocess.rs

```diff
@@ -136,240 +128,238 @@ impl Scope {
-                    ShaderDefValue::Bool(def_value) => {
-                        let val = val.as_str().parse().map_err(|_| {
-                            ComposerErrorInner::InvalidShaderDefComparisonValue {
-                                pos: offset,
-                                shader_def_name: def.as_str().to_string(),
-                                value: val.as_str().to_string(),
-                                expected: "bool".to_string(),
-                            }
-                        })?;
-                        act_on(*def_value, val, op.as_str(), offset)?
-                    }
-                    ShaderDefValue::Int(def_value) => {
-                        let val = val.as_str().parse().map_err(|_| {
-                            ComposerErrorInner::InvalidShaderDefComparisonValue {
-                                pos: offset,
-                                shader_def_name: def.as_str().to_string(),
-                                value: val.as_str().to_string(),
-                                expected: "int".to_string(),
-                            }
-                        })?;
-                        act_on(*def_value, val, op.as_str(), offset)?
-                    }
-                    ShaderDefValue::UInt(def_value) => {
-                        let val = val.as_str().parse().map_err(|_| {
-                            ComposerErrorInner::InvalidShaderDefComparisonValue {
-                                pos: offset,
-                                shader_def_name: def.as_str().to_string(),
-                                value: val.as_str().to_string(),
-                                expected: "int".to_string(),
-                            }
-                        })?;
-                        act_on(*def_value, val, op.as_str(), offset)?
-                    }
-                };
-
-                scope.branch(is_else, new_scope, offset)?;
-            } else if self.else_regex.is_match(line) {
-                scope.branch(true, true, offset)?;
-            } else if self.endif_regex.is_match(line) {
-                scope.pop(offset)?;
-            } else if let Some(cap) = self.define_import_path_regex.captures(line) {
-                name = Some(cap.get(1).unwrap().as_str().to_string());
-            } else if let Some(cap) = self.define_shader_def_regex.captures(line) {
-                if at_start {
-                    still_at_start = true;
+                    loop {
+                        // output spaces for removed lines to keep spans consistent (errors report against substituted_source, which is not preprocessed)
+                        final_string.extend(std::iter::repeat(" ").take(line.len()));
+                        offset += line.len() + 1;
 
-                    let def = cap.get(1).unwrap();
-                    let name = def.as_str().to_string();
+                        open_count += line.match_indices('{').count();
```

Comment:
We're doing multiple passes over `line` here to count open/close curly braces. Allocating a new "final string" for the import lines also feels suboptimal. Not worth blocking on, but I think its worth adding a `PERF` comment here that this could be optimized with a custom impl. I think the ideal impl would find the offset index of the current line being considered in the original "unsplit" shader string, pass in a slice of that string starting at that offset, and run our custom parse logic (ex: the parse_imports code) on that slice (ending on a particular index). We could then resume line-by-line parsing by calling `lines()` on a string slice starting at the index where parsing finished.

Reply:
comment added - it's a shame but we would need to de-comment the whole shader upfront to use a substring ref. that's probably fine but i'd rather do it after this is in.

Reply:
Works for me! It feels like we _should_ be able to implement the preprocessor as a single pass over the input string that produces the final output string. This would largely involve swapping out line-driven regex (and the CommentReplaceIter) for a custom unified parser.